### PR TITLE
Add migration rule and lit test for cmake migration

### PR DESCRIPTION
**Pull request by the-slow-one**
Add rules and lit test for the following cases.

- [x] `CMAKE_C/CXX_STANDARD`
- [x] `set_target_properties`
- [ ] User defined functions
- [x] `CMAKE_C/CXX_COMPILER`
- [ ] `cuda_compile`

This PR resolves [CTST-5674](https://jira.devtools.intel.com/browse/CTST-5674)

Note: The PR is a replacement for [PR-1470](https://github.com/oneapi-src/SYCLomatic/pull/1470/files). The reason for replacement is unexpected compilations of branching. All of issue are resolved in this PR. Also, All comments in PR-1470 are addressed in this PR.

### CODE REVIEW

File: clang/test/dpct/cmake_migation/case_014/expected.txt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# Using magic number to be diverse
+set (CMAKE_CXX_STANDARD 17)
+
+# In the following cases check for NO changes after running the
+# pattern-rewritter. These test will fail due to limitation in
+# pattern-rewritter.
+# set (CMAKE_CXX_STANDARD CPP_VER_17)
+# set (CMAKE_CXX_STANDARD CPP_VER_20)
+# set (CMAKE_CXX_STANDARD CPP_VER_23)
+# set (CMAKE_CXX_STANDARD CPP_VER_26)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_014/expected.txt]: You can migrate "CMAKE_CXX_STANDARD" to "CPP_VER_17" for line20 to line23 for current migration yaml based rule, and add comments for this rule just for temporary use, and an implicit rule for CMAKE_CXX_STANDARD will be added later on.

File: clang/test/dpct/cmake_migation/case_015/expected.txt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+# Test C++ 99 standard
+set_target_properties(target_one PROPERTIES CXX_STANDARD 17)
+
+# Test C++ 99 standard
+set_target_properties(target_one PROPERTIES CXX_STANDARD 17)
+
+# No change unless property matches exactly
+set_target_properties(target_one
+                      PROPERTIES
+                      CXX_Standard 14)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_015/expected.txt:17]: Is CXX_Standard a invalid cmake syntax?

File: clang/test/dpct/cmake_migation/case_014/input.cmake [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+# Using magic number to be diverse
+set (CMAKE_CXX_STANDARD 14)
+
+# In the following cases check for NO changes after running the
+# pattern-rewritter. These test will fail due to limitation in
+# pattern-rewritter.
+# set (CMAKE_CXX_STANDARD CPP_VER_17)
+# set (CMAKE_CXX_STANDARD CPP_VER_20)
+# set (CMAKE_CXX_STANDARD CPP_VER_23)
+# set (CMAKE_CXX_STANDARD CPP_VER_26)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_014/input.cmake]: You can migrate "CMAKE_CXX_STANDARD" to "CPP_VER_17" for line20 to line23 for current migration yaml based rule, and add comments for this rule just for temporary use, and an implicit rule for CMAKE_CXX_STANDARD will be added later on.

File: clang/test/dpct/cmake_migation/case_015/expected.txt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+# No change unless value is one of 99, 11, 14
+# This negetive test will fail due to limitation in pattern-rewritter
+# set_target_properties(target_one
+#                       PROPERTIES
+#                       CXX_STANDARD 20)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_015/expected.txt]: You can migrate "CXX_STANDARD" to "17" for current migration yaml based rule, and add comments for this rule just for temporary use, and an implicit rule for set_target_properties will be added later on.

File: clang/test/dpct/cmake_migation/case_015/expected.txt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+# Test CUDA_SEPARABLE_COMPILATION
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_015/expected.txt:26]: I think make the original cuda syntax commented more friendly than being removed in the migrated code.

File: clang/test/dpct/cmake_migation/case_015/expected.txt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+
+# No change unless property matches exactly
+set_target_properties(target_one
+                      PROPERTIES
+                      Cuda_SEPARABLE_COMPILATION Off)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_015/expected.txt:34]: Cuda_SEPARABLE_COMPILATION is valid syntax in cmake?

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 7]
```diff
@@ -41,3 +41,51 @@
...
+  CmakeSyntax: set_property_cxx_standard
+  In: set_target_properties(${targets} PROPERTIES CXX_STANDARD ${version})
+  Out: set_target_properties(${targets} PROPERTIES CXX_STANDARD 17)
+
+- Rule: rule_set_property_cuda_seprarable_only
+  Kind: CMakeRule
+  Priority: Fallback
+  CmakeSyntax: set_property_cuda_seprarable_only
+  In: set_target_properties(${target_property} CUDA_SEPARABLE_COMPILATION ${state})
+  Out: ""
```
Comment (tomflinda) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml]: For line 70: suggest to use 
```
Out: |
    #[[
      set_target_properties(${target_property} CUDA_SEPARABLE_COMPILATION ${state})
    ]]
```
which will comment the cuda syntax in the migrated script, which is more user friendly than  being removed.

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 8]
```diff
@@ -41,3 +41,51 @@
...
+  CmakeSyntax: set_property_cuda_architechtures
+  In: set_target_properties(${targets} PROPERTIES CUDA_ARCHITECHTURES ${state} ${other_properties})
+  Out: set_target_properties(${targets} PROPERTIES ${other_properties})
+
+- Rule: rule_set_property_cuda_architechtures_only
+  Kind: CMakeRule
+  Priority: Fallback
+  CmakeSyntax: set_property_cuda_architechtures_only
+  In: set_target_properties(${target_property} CUDA_ARCHITECHTURES ${state})
+  Out: ""
```
Comment (tomflinda) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml]: Same as above

File: clang/test/dpct/cmake_migation/case_014/case_014_cmake_cxx_standard.cpp [hunk 9]
```diff
@@ -0,0 +1,11 @@
+// UNSUPPORTED: system-windows
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migation/case_014/case_014_cmake_cxx_standard.cpp]: Line 1 should be removed.

File: clang/test/dpct/cmake_migation/case_015/expected.txt (hunk 2 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_015/expected.txt:17]: `CXX_Standard` is not a CMake defined variable.

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml (hunk 7 above)
Comment (the-slow-one) [reply clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml]: Commenting out the code is much better! Using the comment format as mentioned here

https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#bracket-comment

File: clang/test/dpct/cmake_migation/case_015/input.cmake [hunk 10]
```diff
@@ -0,0 +1,63 @@
+cmake_minimum_required(VERSION 3.24)
+project(AProject)
+cuda_add_library(target_one MODULE nosrc1.cu)
+
+# Test C++ 99 standard
+set_target_properties(target_one
+                      PROPERTIES
+                      CXX_STANDARD 99)
+
+# Test C++ 99 standard
```
Comment (zhiweij1) [inline clang/test/dpct/cmake_migation/case_015/input.cmake]:
```suggestion
# Test C++ 11 standard
```

File: clang/test/dpct/cmake_migation/case_015/input.cmake [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+set_target_properties(target_one
+                      PROPERTIES
+                      CXX_STANDARD 11)
+
+# Test C++ 99 standard
```
Comment (zhiweij1) [inline clang/test/dpct/cmake_migation/case_015/input.cmake]:
```suggestion
# Test C++ 14 standard
```

File: clang/test/dpct/cmake_migation/case_017/expected.txt [hunk 12]
```diff
@@ -0,0 +1,26 @@
+set (GCC_COMPILER g++)
+
+# In the following cases check for changes after running the
+# pattern-rewritter
+set (CMAKE_CXX_COMPILER icpx)
+set (CMAKE_CXX_COMPILER icpx)
+set (CMAKE_CXX_COMPILER icpx)
+set (CMAKE_CXX_COMPILER icpx)
```
Comment (zhiweij1) [inline clang/test/dpct/cmake_migation/case_017/expected.txt:8]: Is it expected to migrate `no_a_cpp_compiler` to `icpx`?

File: clang/test/dpct/cmake_migation/case_015/case_015_set_target_properties.cpp [hunk 13]
```diff
@@ -0,0 +1,11 @@
+// UNSUPPORTED: system-windows
```
Comment (zhiweij1) [inline clang/test/dpct/cmake_migation/case_015/case_015_set_target_properties.cpp]:
```suggestion
```

File: clang/test/dpct/cmake_migation/case_017/case_017_cmake_cxx_compiler.cpp [hunk 14]
```diff
@@ -0,0 +1,11 @@
+// UNSUPPORTED: system-windows
```
Comment (zhiweij1) [inline clang/test/dpct/cmake_migation/case_017/case_017_cmake_cxx_compiler.cpp]:
```suggestion
```

File: clang/test/dpct/cmake_migation/case_014/expected.txt (hunk 1 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_014/expected.txt]: Done

File: clang/test/dpct/cmake_migation/case_014/input.cmake (hunk 3 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_014/input.cmake]: Done

File: clang/test/dpct/cmake_migation/case_015/expected.txt (hunk 4 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_015/expected.txt]: Done

File: clang/test/dpct/cmake_migation/case_015/expected.txt (hunk 5 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_015/expected.txt:26]: Done

File: clang/test/dpct/cmake_migation/case_015/expected.txt (hunk 6 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_015/expected.txt:34]: `Cuda_SEPARABLE_COMPILATION` is not a CMake defined variable.

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml (hunk 8 above)
Comment (the-slow-one) [reply clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml]: Done

File: clang/test/dpct/cmake_migation/case_014/case_014_cmake_cxx_standard.cpp (hunk 9 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_014/case_014_cmake_cxx_standard.cpp]: Done

File: clang/test/dpct/cmake_migation/case_017/case_017_cmake_cxx_compiler.cpp (hunk 14 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_017/case_017_cmake_cxx_compiler.cpp]: Done

File: clang/test/dpct/cmake_migation/case_015/case_015_set_target_properties.cpp (hunk 13 above)
Comment (the-slow-one) [reply clang/test/dpct/cmake_migation/case_015/case_015_set_target_properties.cpp]: Done

**Review by zhiweij1 — APPROVED** [review]

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by tomflinda — COMMENTED** [review]
@DeepakRajHR  you PR need to be rebased to fix conflicts manually

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Comment by zhimingwang36** [comment]
there is conflict in clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml , please resolve the conflicts. Thanks.

**Pull request merged**